Add support for RX8130 RTC Chip

Beormund opened the pull request:
# What does this implement/fix?

<!-- Quick description and explanation of changes -->
Support for the Epson RTC Chip RX8130 used by recent M5Stack devices such as the M5StamPLC controller.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Other

Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):

- https://github.com/esphome/esphome-docs/pull/5315

## Test Environment

- [x] ESP32
- [x] ESP32 IDF
- [x] ESP8266
- [x] RP2040
- [ ] BK72xx
- [ ] RTL87xx
- [ ] nRF52840

## Example entry for `config.yaml`:

```yaml
esphome:
  name:  m5stamplc
  friendly_name: 'M5Stack STAMPLC'
  on_boot:
    - then:
        # Read time from RTC
        rx8130.read_time:
time:
  - platform: rx8130
    id: rx8130_time
    update_interval: 30 min
  - platform: sntp
    id: sntp_time
    timezone: Europe/London
    on_time_sync:
      then:
        # Write time to RTC
        - rx8130.write_time:
            id: rx8130_time
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

Beormund committed:
Add support for RX8130 RTC Chip

Beormund committed:
Merge branch 'esphome:dev' into rx8130

Beormund committed:
Fixes to clear clang-tidy checks

Updated CODEOWNERS and implemented fixes for clang-tidy errors.

Beormund committed:
Merge branch 'rx8130' of https://github.com/Beormund/esphome into rx8130

Beormund committed:
Merge branch 'dev' into rx8130

Beormund committed:
Merge branch 'esphome:dev' into rx8130

Beormund committed:
Added Tests for RX8130 Time Component

jesserockz committed:
Merge branch 'dev' into rx8130

jesserockz commented on esphome/components/rx8130/rx8130.cpp:
```diff
@@ -0,0 +1,125 @@
+static const char *const TAG = "rx8130";
+
+void RX8130Component::setup() {
+  ESP_LOGCONFIG(TAG, "Setting up RX8130...");
```
```suggestion
```

Beormund replied:
Done

jesserockz commented on esphome/components/rx8130/rx8130.cpp:
```diff
@@ -0,0 +1,125 @@
+  ESP_LOGCONFIG(TAG, "  FLAGS  %s", format_bin((uint8_t) this->reg(RX8130_REG_FLAG)).c_str());
+  ESP_LOGCONFIG(TAG, "  CTRL0  %s", format_bin((uint8_t) this->reg(RX8130_REG_CTRL0)).c_str());
+  ESP_LOGCONFIG(TAG, "  CTRL1  %s", format_bin((uint8_t) this->reg(RX8130_REG_CTRL1)).c_str());
+  ESP_LOGCONFIG(TAG, "  Timezone: '%s'", this->timezone_.c_str());
```
Please combine these logs: https://developers.esphome.io/architecture/logging/#best-practices-for-combined-logging

jesserockz replied:
Also actually, dump_config is not supposed to do any reads etc, it is only supposed to dump the stuff that is already known, or was configured by the user.

Beormund replied:
Removed logs with reads.

jesserockz commented on esphome/components/rx8130/rx8130.cpp:
```diff
@@ -0,0 +1,125 @@
+  if (this->is_failed()) {
+    ESP_LOGE(TAG, "Communication with RX8130 failed!");
+    return;
+  }
```
Not required as the base component will log this
```suggestion
```

Beormund replied:
Done

jesserockz commented on esphome/components/rx8130/rx8130.cpp:
```diff
@@ -0,0 +1,125 @@
+}
+
+float RX8130Component::get_setup_priority() const { return setup_priority::DATA; }
+
```
No need to override when using the default value
```suggestion
```

Beormund replied:
Removed

jesserockz commented on esphome/components/rx8130/rx8130.cpp:
```diff
@@ -0,0 +1,125 @@
+
+uint8_t RX8130Component::bcd2dec_(uint8_t val) { return (val >> 4) * 10 + (val & 0x0f); }
+
+uint8_t RX8130Component::dec2bcd_(uint8_t val) { return ((val / 10) << 4) + (val % 10); }
```
These should be moved to the top of the file and made into constexpr. They dont need to be class members

```suggestion
constexpr uint8_t bcd2dec(uint8_t val) { return (val >> 4) * 10 + (val & 0x0f); }

constexpr uint8_t dec2bcd(uint8_t val) { return ((val / 10) << 4) + (val % 10); }
```

Beormund replied:
Done

jesserockz commented on esphome/components/rx8130/rx8130.cpp:
```diff
@@ -0,0 +1,125 @@
+  uint8_t date[7];
+  if (this->read_register(RX8130_REG_SEC, date, 7) != ERROR_OK) {
+    ESP_LOGE(TAG, "Can't read I2C data.");
+    this->mark_failed();
```
We shouldn't fail the component if we failed to read once. Maybe use a `set_retry` to try multiple times before actually failing it?
```suggestion
    this->status_set_warning(ESP_LOG_MSG_COMM_FAIL);
```

Beormund replied:
Removed

jesserockz commented on esphome/components/rx8130/rx8130.cpp:
```diff
@@ -0,0 +1,125 @@
+  this->stop_(true);
+  if (this->write_register(RX8130_REG_SEC, buff, 7) != ERROR_OK) {
+    ESP_LOGE(TAG, "Can't write I2C data.");
+    this->mark_failed();
```
Same here, dont mark failed, try again next time (or retry)

Beormund replied:
Removed

jesserockz commented on esphome/components/rx8130/rx8130.h:
```diff
@@ -0,0 +1,50 @@
+ protected:
+  void stop_(bool stop);
+  uint8_t bcd2dec_(uint8_t val);
+  uint8_t dec2bcd_(uint8_t val);
```
```suggestion
```

Beormund replied:
Done

Beormund committed:
Merge branch 'esphome:dev' into rx8130

Beormund committed:
Implemented suggested changes to rx8130.cpp

Tidied up/combined logging per best practices; removed get_setup_priority to use default; converted bcd2dec/dec2bcd class members to constexpr; removed mark_failed() to allow retry next time.

Beormund committed:
Removed i2c reads in dump_config()

lboue commented:
It works fine with M5Tab:

**Logs**
```
[21:01:00][D][rx8130:077]: Read UTC time: 2025-10-22 19:00:51
[21:01:00][D][time:061]: Synchronized time: 2025-10-22 19:00:51
[21:01:40][D][sntp:099]: Synchronized time: 2025-10-22 19:01:31
[21:01:40][D][rx8130:100]: Wrote UTC time: 2025-10-22 19:01:31
...
[21:16:40][D][sntp:099]: Synchronized time: 2025-10-22 19:16:31
[21:16:40][D][rx8130:100]: Wrote UTC time: 2025-10-22 19:16:31
```

**Config**
```yaml
time:
  # https://docs.m5stack.com/en/arduino/m5tab5/rtc
  - platform: rx8130
    id: rx8130_time
    # RX8130CE (0x32)
    update_interval: 30 min
  - platform: sntp
    id: sntp_time
    timezone: Europe/Paris
    on_time_sync:
      then:
        # Write time to RTC
        - rx8130.write_time:
            id: rx8130_time
```

swoboda1337 commented on tests/components/rx8130/common.yaml:
Tests need to be updated after the common bus changes. Thanks!

Beormund replied:
I don't understand. What do I need to update?

swoboda1337 replied:
Take a look at other i2c tests:
https://github.com/esphome/esphome/blob/dev/tests/components/am2315c/test.esp32-c3-idf.yaml
https://github.com/esphome/esphome/blob/dev/tests/components/am2315c/common.yaml

swoboda1337 replied:
Can also remove the esp32 ard tests. Thanks!

Beormund replied:
The links you provided above are broken and I can't find the paths to the include files. What "esp32 and tests" are you referring to. All very confusing.

swoboda1337 replied:
Sorry someone remove the test I linked to take a look here: 
https://github.com/esphome/esphome/blob/dev/tests/components/am2315c/

swoboda1337 replied:
Modify your tests files to match that.

lboue commented on tests/components/rx8130/common.yaml:
```diff
@@ -0,0 +1,12 @@
+    - rx8130.read_time
+    - rx8130.write_time
+
+i2c:
+  - id: i2c_rx8130
+    scl: ${scl_pin}
+    sda: ${sda_pin}
+
+time:
+  - platform: rx8130
```
```suggestion
esphome:
  on_boot:
    - rx8130.read_time
    - rx8130.write_time

time:
  - platform: rx8130
    i2c_id: i2c_bus
```

lboue commented on tests/components/rx8130/test.esp32-c3-idf.yaml:
```diff
@@ -0,0 +1,5 @@
+  scl_pin: GPIO5
+  sda_pin: GPIO4
+
+<<: !include common.yaml
```
```suggestion
packages:
  i2c: !include ../../test_build_components/common/i2c/esp32-idf.yaml

<<: !include common.yaml
```

lboue commented on tests/components/rx8130/test.rp2040-ard.yaml:
```diff
@@ -0,0 +1,5 @@
+  scl_pin: GPIO5
+  sda_pin: GPIO4
+
+<<: !include common.yaml
```
```suggestion
packages:
  i2c: !include ../../test_build_components/common/i2c/rp2040-ard.yaml

<<: !include common.yaml
```

lboue commented on tests/components/rx8130/test.esp32-idf.yaml:
```diff
@@ -0,0 +1,5 @@
+  scl_pin: GPIO16
+  sda_pin: GPIO17
+
+<<: !include common.yaml
```
```suggestion
packages:
  i2c: !include ../../test_build_components/common/i2c/esp32-idf.yaml

<<: !include common.yaml
```

lboue commented on tests/components/rx8130/test.esp32-c3-ard.yaml:
```diff
@@ -0,0 +1,5 @@
+  scl_pin: GPIO5
+  sda_pin: GPIO4
+
+<<: !include common.yaml
```
Remove this file:
```suggestion
```

Beormund committed:
Updated rx8130 tests

Beormund committed:
Merge branch 'dev' into rx8130

lboue commented on tests/components/rx8130/test.esp32-c3-ard.yaml:
```diff
@@ -0,0 +1,4 @@
+packages:
+  i2c: !include ../../test_build_components/common/i2c/esp32-ard.yaml
+
+<<: !include common.yaml
```
Please remove this file. The goal is to migration to esp-idf for esp32 boards:
```suggestion
```

Beormund committed:
Remove esp32-c3 tests

lboue commented:
@Beormund Now you can resolve 'outdated' comments and request re-rewiew

swoboda1337 committed:
Delete tests/components/rx8130/test.esp32-ard.yaml

swoboda1337 commented on esphome/components/rx8130/rx8130.h:
```diff
@@ -0,0 +1,47 @@
+static const uint8_t RX8130_REG_CTRL0 = 0x1E;
+static const uint8_t RX8130_REG_CTRL1 = 0x1F;
+static const uint8_t RX8130_REG_DIG_OFFSET = 0x30;
+static const uint8_t RX8130_BIT_CTRL_STOP = 0x40;
```
Could move to the cpp file as they are not used in header.

lboue replied:
```suggestion
```

swoboda1337 commented on esphome/components/rx8130/rx8130.cpp:
```diff
@@ -0,0 +1,109 @@
+
+namespace esphome {
+namespace rx8130 {
+using namespace i2c;
```
Would be better to use `i2c::` in the few places that need it, if any.

swoboda1337 commented on esphome/components/rx8130/rx8130.cpp:
```diff
@@ -0,0 +1,109 @@
+  this->stop_(false);
+}
+
+void RX8130Component::stop_(bool stop) { this->reg(RX8130_REG_CTRL0) = stop ? RX8130_BIT_CTRL_STOP : 0x00; }
```
No error check here

swoboda1337 commented on esphome/components/rx8130/rx8130.cpp:
```diff
@@ -0,0 +1,109 @@
+    return;
+  }
+  // Enable battery charging and switching
+  uint8_t battery_flags = 0x30;
```
Could add a constant for this.

swoboda1337 reviewed (changes requested):
A few more minor things otherwise it looks good.

lboue commented on esphome/components/rx8130/rx8130.cpp:
```diff
@@ -0,0 +1,109 @@
+    return;
+  }
+  // Disable wakeup timers
+  if (this->write_register(RX8130_REG_EXTEN, &clear_flags, 1) != ERROR_OK) {
```
```suggestion
  if (this->write_register(RX8130_REG_EXTEN, &clear_flags, 1) != i2c::ERROR_OK) {
```

lboue commented on esphome/components/rx8130/rx8130.cpp:
```diff
@@ -0,0 +1,109 @@
+    return;
+  }
+  // Clear test flag and disable interrupts
+  if (this->write_register(RX8130_REG_CTRL0, &clear_flags, 1) != ERROR_OK) {
```
```suggestion
  if (this->write_register(RX8130_REG_CTRL0, &clear_flags, 1) != i2c::ERROR_OK) {
```

lboue commented on esphome/components/rx8130/rx8130.cpp:
```diff
@@ -0,0 +1,109 @@
+  buff[5] = dec2bcd(now.month);
+  buff[6] = dec2bcd(now.year % 100);
+  this->stop_(true);
+  if (this->write_register(RX8130_REG_SEC, buff, 7) != ERROR_OK) {
```
```suggestion
  if (this->write_register(RX8130_REG_SEC, buff, 7) != i2c::ERROR_OK) {
```

lboue commented on esphome/components/rx8130/rx8130.h:
```diff
@@ -0,0 +1,47 @@
+  void update() override;
+  void dump_config() override;
+  void read_time();
+  void write_time();
```
```suggestion
  void write_time();
  /// Ensure RTC is initialized at the correct time in the setup sequence
  float get_setup_priority() const override { return setup_priority::DATA; }
```

Beormund committed:
Tidied up constants and implemented recommended copilot changes

Beormund committed:
Merge branch 'dev' into rx8130

Beormund committed:
Update rx8130.cpp

Missing closing brace after removal of i2c namespace

Beormund committed:
Merge branch 'rx8130' of https://github.com/Beormund/esphome into rx8130

swoboda1337 commented on esphome/components/rx8130/rx8130.cpp:
```diff
@@ -0,0 +1,126 @@
+}
+
+void RX8130Component::stop_(bool stop) {
+  if (this->write_register(RX8130_REG_CTRL0, stop ? &RX8130_BIT_CTRL_STOP : &RX8130_CLEAR_FLAGS, 1) != i2c::ERROR_OK) {
```
The `&` doesn't make any sense here. You are writing the esp32 address of those constants not the constants themselves.

Beormund replied:
Might be best if I close this PR and leave it to someone who know what they are doing, or I come back to it in the future when my knowledge of C++ has improved. Cheers.

Beormund replied:
What is the write_register equivalent of this->reg(RX8130_REG_CTRL0) = RX8130_BIT_CTRL_STOP ?

write_register expects a pointer to an array for the data arg. RX8130_BIT_CTRL_STOP is a uint8_t. Do I need to redefine it as RX8130_BIT_CTRL_STOP[8] = {0x00}?

swoboda1337 replied:
Apologies it was correct. It just looked weird. Have been staring at diffs for too many hours today :)

swoboda1337 replied:
It looked odd with the ternary operator I should have looked more carefully. Sorry about that

swoboda1337 commented on esphome/components/rx8130/rx8130.cpp:
```diff
@@ -0,0 +1,126 @@
+}
+
+void RX8130Component::stop_(bool stop) {
+  if (this->write_register(RX8130_REG_CTRL0, stop ? &RX8130_BIT_CTRL_STOP : &RX8130_CLEAR_FLAGS, 1) != i2c::ERROR_OK) {
```
```suggestion
  const uint8_t data = stop ? RX8130_BIT_CTRL_STOP : RX8130_CLEAR_FLAGS;
  if (this->write_register(RX8130_REG_CTRL0, &data, 1) != i2c::ERROR_OK) {
```

swoboda1337 committed:
Update esphome/components/rx8130/rx8130.cpp

swoboda1337 commented:
Thanks!
